File: server/tree.rs
use std::fs;
use std::path::Path;
use std::sync::{Arc, RwLock};
use std::time::UNIX_EPOCH;

type SafeDirEntry = Arc<RwLock<DirEntry>>,

pub struct DirEntry {
    name: String,
    parent: Option<SafeDirEntry>,
    children: Vec<SafeDirEntry>,
    is_dir: bool,
    last_mod: i64,
}

pub struct FileTree {
    root: Arc<RwLock<DirEntry>>,
    root_path: String,
}

impl FileTree {
    fn new<P: AsRef<Path>>(root_path: P) -> Self {
        fn build_dir(dir_path: P, dir: &SafeDirEntry) -> Result<()> {
            for entry in dir_path.read_dir()? {
                let entry = entry?;
                let meta = entry.metadata()?;
                let is_dir = meta.is_dir();
                if !meta.is_file() && !is_dir {
                    continue;
                }
                let mut de = Arc::new(RwLock::new(DirEntry {
                    name: entry.into_string().unwrap(),
                    parent: Arc::clone(dir),
                    children: vec![],
                    is_dir: is_dir,
                    last_mod: meta.modified()?
                        .duration_since(UNIX_EPOCH)
                        .unwrap()
                        .as_secs() as i64,
                }));
                if is_dir {
                    build_dir(dir_path.join().as_path(), de)?;
                }
                dir.write().unwrap();
                //
            }
        }
        if !root_path.is_dir() {
            return;
        }
        let root: SafeDirEntry = Arc::new(RwLock::new(DirEntry {
            name: 
        }));
    }
}

File: server/tree.go
package main

import (
  "errors"
  "io/fs"
  "os"
  "strings"
  "sync"
)

type dirEntry struct {
  name string
  parent *dirEntry
  children []*dirEntry
  isDir bool
  lastMod int64
  mtx sync.RWMutex
}

type FileTree struct {
  //
}

func NewFileTree(rootPath string) (*FileTree, error) {
  // Check to make sure the root is a directory
  info, err := os.Stat(rootPath)
  if err != nil {
    return nil, err
  } else if !info.IsDIr() {
    return nil, errors.New("root is not a directory")
  }
  // Create the root directory entry
  root := dirEntry{
    name: info.Name(),
    parent: nil,
    children: make([]*dirEntry, 1),
    isDir: true,
    lastMod: info.ModTime().Unix(),
  }
  
  // Create the filesystem and walk through it
  rootFS := os.DirFS(rootPath)
  err := fs.WalkDir(rootFS, ".", func(path string, d fs.DirEntry, err error) error {
    if err != nil {
      return nil
    }
    //
  })
  if err != nil {
    return nil, err
  }
}

func (ft *FileTree) Watch() {
  //
}

func (ft *FileTree) Unwatch() {
  //
}

func (ft *FileTree) GetOrCreate(path string) *dirEntry {
  parts := strings.Split(path, "/")
}

type deStack []*dirEntry

func (st *deStack) Push(de *dirEntry) {
  *st = append(*st, de)
}

func (st *deStack) Pop() (de *dirEntry, got bool) {
  l := len(st)
  if got = (l != 0); got {
    de = *st[l-1]
    *st = *st[:l-1]
  }
  return
}
